src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -321,23 +326,29 @@ private static unsafe long SendFile(SafeSocketHandle socket, SafeFileHandle file
 
         private static unsafe int SysReceive(SafeSocketHandle socket, SocketFlags flags, IList<ArraySegment<byte>> buffers, byte[]? socketAddress, ref int socketAddressLen, out SocketFlags receivedFlags, out Interop.Error errno)
         {
-            int available = 0;
-            errno = Interop.Sys.GetBytesAvailable(socket, &available);
-            if (errno != Interop.Error.SUCCESS)
-            {
-                receivedFlags = 0;
-                return -1;
-            }
-            if (available == 0)
+            int maxBuffers = buffers.Count;
+            bool allocOnStack = maxBuffers <= IovStackThreshold;
+
+            // When there are many buffers, reduce the number of pinned buffers based on available bytes.
+            int available = int.MaxValue;
+            if (!allocOnStack)
             {
-                // Don't truncate iovecs.
-                available = int.MaxValue;
+                errno = Interop.Sys.GetBytesAvailable(socket, &available);
+                if (errno != Interop.Error.SUCCESS)
+                {
+                    receivedFlags = 0;
+                    return -1;
+                }
+                if (available == 0)
+                {
+                    // Don't truncate iovecs.
+                    available = int.MaxValue;
+                }
             }
 
             // Pin buffers and set up iovecs.
-            int maxBuffers = buffers.Count;
-            var handles = new GCHandle[maxBuffers];
-            var iovecs = new Interop.Sys.IOVector[maxBuffers];
+            Span<GCHandle> handles = allocOnStack ? stackalloc GCHandle[IovStackThreshold] : new GCHandle[maxBuffers];
+            Span<Interop.Sys.IOVector> iovecs = allocOnStack ? stackalloc Interop.Sys.IOVector[IovStackThreshold] : new Interop.Sys.IOVector[maxBuffers];
```

Comment:
The Windows implementation uses ArrayPool as a fallback:
https://github.com/dotnet/runtime/blob/84abb55039eb7d94b435cd79104901b25ed1e2bd/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs#L359
That might be overkill.  I don't have a good sense for how many buffers someone is realistically going to provide and thus how often this path would actually be hit.

Reply:
I suspect that having lots of buffers (e.g. more than 4) is quite rare.

Reply:
I suspect the same thing, and the 'low' value of `UIO_FASTIOV` suggests this also.